src/phoenix/db/insertion/evaluation.py

```diff
@@ -0,0 +1,169 @@
+from typing import NamedTuple, Optional
+
+from sqlalchemy import insert, select
+from sqlalchemy.ext.asyncio import AsyncSession
+from typing_extensions import assert_never
+
+from phoenix.db import models
+from phoenix.db.helpers import SupportedSQLDialect, num_docs_col
+from phoenix.exceptions import PhoenixException
+from phoenix.trace import v1 as pb
+
+
+class InsertEvaluationError(PhoenixException):
+    pass
+
+
+class _EvaluationInsertionResult(NamedTuple):
+    project_rowid: int
+    evaluation_name: str
+
+
+class SpanEvaluationInsertionResult(_EvaluationInsertionResult): ...
+
+
+class TraceEvaluationInsertionResult(_EvaluationInsertionResult): ...
+
+
+class DocumentEvaluationInsertionResult(_EvaluationInsertionResult): ...
+
+
+async def insert_evaluation(
+    session: AsyncSession,
+    evaluation: pb.Evaluation,
+) -> Optional[_EvaluationInsertionResult]:
+    evaluation_name = evaluation.name
+    result = evaluation.result
+    label = result.label.value if result.HasField("label") else None
+    score = result.score.value if result.HasField("score") else None
+    explanation = result.explanation.value if result.HasField("explanation") else None
+    if (evaluation_kind := evaluation.subject_id.WhichOneof("kind")) is None:
+        raise InsertEvaluationError("Cannot insert an evaluation that has no evaluation kind")
+    elif evaluation_kind == "trace_id":
+        trace_id = evaluation.subject_id.trace_id
+        return await _insert_trace_evaluation(
+            session, trace_id, evaluation_name, label, score, explanation
+        )
+    elif evaluation_kind == "span_id":
+        span_id = evaluation.subject_id.span_id
+        return await _insert_span_evaluation(
+            session, span_id, evaluation_name, label, score, explanation
+        )
+    elif evaluation_kind == "document_retrieval_id":
+        span_id = evaluation.subject_id.document_retrieval_id.span_id
+        document_position = evaluation.subject_id.document_retrieval_id.document_position
+        return await _insert_document_evaluation(
+            session, span_id, document_position, evaluation_name, label, score, explanation
+        )
+    else:
+        assert_never(evaluation_kind)
```

Comment:
Curious if there are any utils in `functools` that have dispatch patterns that we could use in this case where we want to dispatch based on a derived value and not a type.

Reply:
since it's not a dispatch based on type, this is probably how it needs to be done